# SQL Part 3 Exercises Joins

This notebook works with the BooksDB dataset.  

**Questions 1-4** focus on how joins work in a more general sense.  If you are not sure how to best answer the questions, try manipulating the queries by changing the order of joins, the length of the TOP, or how they are ordered.  

> <span style="background-color: rgba(127, 127, 127, 0.1);"><i>Note</i>: If some of the queries are not returning quickly, you can try closing and restarting azure data studio after a few minutes.</span>

In **questions 5-8**, you will be asked to join tables to find results about titles, tags, and the books users want to read.

As you go through the questions, you will see 'Minimum Desired Output'.  This will provide you with the bare minimum columns needed to answer the questions.  When you write the queries, add as many additional columns as you need to best find your answers.

## **QUESTION 1:**  Exploring a LEFT JOIN using the **books** and **to\_read** tables

The code block below has been commented out. To answer this question, you will need to uncomment it by removing the **/\*** (line 1) and **\*/** (line 16).

**Part A:** Look at Query 1, which table is the 'Left' table and which is the 'Right'?

> Your Answer:  books is the 'Left' table and to\_read is the 'Right' table.

**Part B:** Using table 2, answer the following questions: 

- When we added a  '**WHERE tr.user\_id IS NOT NULL'** clause (line 14), how did the table change?  Why?

> Your Answer:  It filtered/removed the NULL value rows from the output.

In [1]:

-- Query 1: LEFT JOIN 
SELECT TOP 50 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
LEFT JOIN BooksDB.dbo.to_read AS tr
ON b.best_book_id = tr.book_id
ORDER BY b.average_rating;

-- Query 2: LEFT JOIN USING 'IS NOT NULL'
SELECT TOP 50 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
LEFT JOIN BooksDB.dbo.to_read AS tr
ON b.best_book_id = tr.book_id
WHERE tr.user_id IS NOT NULL
ORDER BY b.average_rating;

(50 rows affected)

(50 rows affected)

Total execution time: 00:00:07.586

title average_rating books_count user_id book_id One Night at the Call Center 2.47 25 NULL NULL The Almost Moon 2.67 63 NULL NULL The Finkler Question 2.76 45 NULL NULL Four Blondes 2.8 57 449 6613 Four Blondes 2.8 57 1378 6613 Four Blondes 2.8 57 6465 6613 Four Blondes 2.8 57 6700 6613 Four Blondes 2.8 57 7830 6613 Four Blondes 2.8 57 10085 6613 Four Blondes 2.8 57 16298 6613 Four Blondes 2.8 57 16845 6613 Four Blondes 2.8 57 17470 6613 Four Blondes 2.8 57 19763 6613 Four Blondes 2.8 57 21938 6613 Four Blondes 2.8 57 22289 6613 Four Blondes 2.8 57 23758 6613 Four Blondes 2.8 57 27874 6613 Four Blondes 2.8 57 28690 6613 Four Blondes 2.8 57 29023 6613 Four Blondes 2.8 57 29315 6613 Four Blondes 2.8 57 30245 6613 Four Blondes 2.8 57 36128 6613 Four Blondes 2.8 57 36554 6613 Four Blondes 2.8 57 38159 6613 Four Blondes 2.8 57 38800 6613 Four Blondes 2.8 57 40361 6613 Four Blondes 2.8 57 40529 6613 Four Blondes 2.8 57 41485 6613 Four Blondes 2.8 57 42706 6613 Four Blondes 2.8 57 45206 6613 Four Blondes 2.8 57 45689 6613 Four Blondes 2.8 57 49499 6613 Four Blondes 2.8 57 51105 6613 Four Blondes 2.8 57 51671 6613 Four Blondes 2.8 57 52109 6613 Lost 2.8 18 NULL NULL Tell-All 2.84 37 NULL NULL Revenge Wears Prada: The Devil Returns (The Devil Wears Prada, #2) 2.84 59 NULL NULL Among the Ten Thousand Things 2.84 19 NULL NULL The Emperor's Children 2.93 45 7860 3648 The Emperor's Children 2.93 45 11320 3648 The Emperor's Children 2.93 45 12686 3648 The Emperor's Children 2.93 45 12729 3648 The Emperor's Children 2.93 45 12891 3648 The Emperor's Children 2.93 45 16933 3648 The Emperor's Children 2.93 45 17093 3648 The Emperor's Children 2.93 45 24036 3648 The Emperor's Children 2.93 45 35122 3648 The Emperor's Children 2.93 45 36022 3648 The Emperor's Children 2.93 45 36952 3648

title average_rating books_count user_id book_id Four Blondes 2.8 57 449 6613 Four Blondes 2.8 57 1378 6613 Four Blondes 2.8 57 6465 6613 Four Blondes 2.8 57 6700 6613 Four Blondes 2.8 57 7830 6613 Four Blondes 2.8 57 10085 6613 Four Blondes 2.8 57 16298 6613 Four Blondes 2.8 57 16845 6613 Four Blondes 2.8 57 17470 6613 Four Blondes 2.8 57 19763 6613 Four Blondes 2.8 57 21938 6613 Four Blondes 2.8 57 22289 6613 Four Blondes 2.8 57 23758 6613 Four Blondes 2.8 57 27874 6613 Four Blondes 2.8 57 28690 6613 Four Blondes 2.8 57 29023 6613 Four Blondes 2.8 57 29315 6613 Four Blondes 2.8 57 30245 6613 Four Blondes 2.8 57 36128 6613 Four Blondes 2.8 57 36554 6613 Four Blondes 2.8 57 38159 6613 Four Blondes 2.8 57 38800 6613 Four Blondes 2.8 57 40361 6613 Four Blondes 2.8 57 40529 6613 Four Blondes 2.8 57 41485 6613 Four Blondes 2.8 57 42706 6613 Four Blondes 2.8 57 45206 6613 Four Blondes 2.8 57 45689 6613 Four Blondes 2.8 57 49499 6613 Four Blondes 2.8 57 51105 6613 Four Blondes 2.8 57 51671 6613 Four Blondes 2.8 57 52109 6613 The Emperor's Children 2.93 45 7860 3648 The Emperor's Children 2.93 45 11320 3648 The Emperor's Children 2.93 45 12686 3648 The Emperor's Children 2.93 45 12729 3648 The Emperor's Children 2.93 45 12891 3648 The Emperor's Children 2.93 45 16933 3648 The Emperor's Children 2.93 45 17093 3648 The Emperor's Children 2.93 45 24036 3648 The Emperor's Children 2.93 45 35122 3648 The Emperor's Children 2.93 45 36022 3648 The Emperor's Children 2.93 45 36952 3648 The Emperor's Children 2.93 45 38430 3648 The Emperor's Children 2.93 45 39414 3648 The Emperor's Children 2.93 45 45261 3648 The Emperor's Children 2.93 45 45497 3648 The Emperor's Children 2.93 45 45747 3648 The Emperor's Children 2.93 45 48502 3648 The Emperor's Children 2.93 45 49233 3648

## **QUESTION 2:** Exploring **RIGHT** joins with the **books** and **to\_read** tables

You will need to uncomment the Query to answer the question.

**Part A:** Looking at the query, which table is the 'Left' and which is the 'Right'?

> Your Answer:  books is the 'Left' table and to\_read is the 'Right' table.

**Part B:** How do the RIGHT-joined tables differ from the LEFT-joined tables in Question 1?  

> Your Answer: left join - Returns results with all the rows from the books table with null values for unmatched rows from the to\_read table
> 
> right join - Returns results with all the rows from the to\_read table with null values for unmatched rows from the books table

In [3]:

-- Query 1: RIGHT JOIN
SELECT TOP 300 b.title, b.average_rating, b.books_count, tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
RIGHT JOIN BooksDB.dbo.to_read AS tr
ON b.best_book_id = tr.book_id;


(300 rows affected)

Total execution time: 00:00:00.955

title average_rating books_count user_id book_id Children of Dune (Dune Chronicles #3) 3.9 116 1 112 NULL NULL NULL 1 235 NULL NULL NULL 1 533 NULL NULL NULL 1 1198 NULL NULL NULL 1 1874 NULL NULL NULL 1 2058 NULL NULL NULL 1 3334 NULL NULL NULL 2 4 NULL NULL NULL 2 11 The Ultimate Hitchhiker's Guide to the Galaxy 4.37 32 2 13 NULL NULL NULL 2 16 NULL NULL NULL 2 92 Dune Messiah (Dune Chronicles #2) 3.86 140 2 106 NULL NULL NULL 2 113 NULL NULL NULL 2 233 NULL NULL NULL 2 247 NULL NULL NULL 2 380 NULL NULL NULL 2 408 NULL NULL NULL 2 439 NULL NULL NULL 2 479 NULL NULL NULL 2 516 NULL NULL NULL 2 619 NULL NULL NULL 2 1069 The Iliad 3.83 1726 2 1371 NULL NULL NULL 2 1384 NULL NULL NULL 2 1597 NULL NULL NULL 2 1926 NULL NULL NULL 2 1967 NULL NULL NULL 2 1969 NULL NULL NULL 2 2220 NULL NULL NULL 2 2388 NULL NULL NULL 2 3222 NULL NULL NULL 2 3227 NULL NULL NULL 2 3733 NULL NULL NULL 2 5418 NULL NULL NULL 2 6937 NULL NULL NULL 2 7146 NULL NULL NULL 2 7873 NULL NULL NULL 2 8417 Harry Potter Collection (Harry Potter, #1-6) 4.73 6 3 10 NULL NULL NULL 3 66 NULL NULL NULL 3 233 NULL NULL NULL 3 726 NULL NULL NULL 3 932 NULL NULL NULL 3 1146 NULL NULL NULL 3 2439 NULL NULL NULL 3 3566 NULL NULL NULL 3 4582 NULL NULL NULL 3 5448 NULL NULL NULL 3 5765 NULL NULL NULL 3 6274 NULL NULL NULL 3 8403 NULL NULL NULL 5 31 The Virtue of Selfishness: A New Concept of Egoism 3.47 53 5 665 NULL NULL NULL 5 3074 Sherlock Holmes: The Complete Novels and Stories, Volume I 4.45 13 5 3581 NULL NULL NULL 6 64 NULL NULL NULL 6 459 NULL NULL NULL 6 619 NULL NULL NULL 6 649 NULL NULL NULL 6 687 NULL NULL NULL 6 874 The War of Art: Break Through the Blocks & Win Your Inner Creative Battles 4.07 29 6 1319 NULL NULL NULL 6 1912 NULL NULL NULL 6 5269 NULL NULL NULL 6 5537 NULL NULL NULL 6 5820 NULL NULL NULL 6 6220 NULL NULL NULL 6 7480 Harry Potter Collection (Harry Potter, #1-6) 4.73 6 7 10 NULL NULL NULL 7 215 NULL NULL NULL 7 225 NULL NULL NULL 7 233 NULL NULL NULL 7 1322 NULL NULL NULL 7 3552 NULL NULL NULL 7 3720 NULL NULL NULL 7 4468 NULL NULL NULL 8 19 NULL NULL NULL 8 498 NULL NULL NULL 8 1357 NULL NULL NULL 8 2177 NULL NULL NULL 8 3930 The Pilgrimage 3.64 171 8 4004 NULL NULL NULL 8 8792 NULL NULL NULL 8 9304 Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 4.77 11 9 8 Children of Dune (Dune Chronicles #3) 3.9 116 9 112 NULL NULL NULL 9 139 NULL NULL NULL 9 230 NULL NULL NULL 9 233 NULL NULL NULL 9 473 NULL NULL NULL 9 638 NULL NULL NULL 9 653 NULL NULL NULL 9 672 NULL NULL NULL 9 701 NULL NULL NULL 9 736 NULL NULL NULL 9 883 NULL NULL NULL 9 895 NULL NULL NULL 9 1467 NULL NULL NULL 9 1582 NULL NULL NULL 9 1873 NULL NULL NULL 9 2516 NULL NULL NULL 9 2630 NULL NULL NULL 9 3210 Icy Sparks 3.68 27 9 3476 NULL NULL NULL 9 3560 NULL NULL NULL 9 3650 NULL NULL NULL 9 5286 NULL NULL NULL 9 6636 Harry Potter and the Goblet of Fire (Harry Potter, #4) 4.53 332 10 6 NULL NULL NULL 10 385 Atlas Shrugged 3.68 147 10 662 NULL NULL NULL 10 896 NULL NULL NULL 10 1076 NULL NULL NULL 10 1082 NULL NULL NULL 10 1220 NULL NULL NULL 10 1281 NULL NULL NULL 10 1439 NULL NULL NULL 10 1851 NULL NULL NULL 10 1920 NULL NULL NULL 10 2441 NULL NULL NULL 10 2570 NULL NULL NULL 10 2616 NULL NULL NULL 10 3059 NULL NULL NULL 10 3545 NULL NULL NULL 10 3961 NULL NULL NULL 10 4369 NULL NULL NULL 10 4510 NULL NULL NULL 10 5104 NULL NULL NULL 10 5254 NULL NULL NULL 10 5556 NULL NULL NULL 10 6114 NULL NULL NULL 10 6917 NULL NULL NULL 10 7545 NULL NULL NULL 10 7842 NULL NULL NULL 11 3643 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 12 1 NULL NULL NULL 12 124 NULL NULL NULL 12 1471 NULL NULL NULL 12 2893 NULL NULL NULL 13 81 NULL NULL NULL 13 277 NULL NULL NULL 13 318 NULL NULL NULL 13 430 NULL NULL NULL 13 437 NULL NULL NULL 13 551 NULL NULL NULL 13 560 NULL NULL NULL 13 610 NULL NULL NULL 13 678 NULL NULL NULL 13 979 NULL NULL NULL 13 1082 NULL NULL NULL 13 1464 NULL NULL NULL 13 1577 NULL NULL NULL 13 2018 NULL NULL NULL 13 2418 NULL NULL NULL 13 3257 NULL NULL NULL

## **QUESTION 3:** Exploring FULL JOINS with the **books** and **to\_read** tables

You will need to uncomment the Query to answer the question.

**Part A:**  Look at the table and explore how it was populated.  Try manipulating the query to better understand how this table works.  Try adding a WHERE clause, or ORDER BY a column on the **books** table.  Based on your exploration, can you think of when you might want a FULL join rather than a LEFT, RIGHT, or INNER?

> Your Answer: Since we are using full join it got all the values from both tables. Based on my exploration I don't think we need FULL OUTER JOIN as we can get the required details either through left or right join.

In [6]:

-- FULL JOIN
SELECT TOP 30 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
FULL JOIN BooksDB.dbo.to_read AS tr
ON b.book_id = tr.book_id
WHERE b.title IS NOT NULL;

(30 rows affected)

Total execution time: 00:00:00.138

title average_rating books_count user_id book_id Children of Dune (Dune Chronicles #3) 3.9 116 1 112 The Hitchhiker's Guide to the Galaxy (Hitchhiker's Guide to the Galaxy, #1) 4.2 257 2 11 The Ultimate Hitchhiker's Guide to the Galaxy 4.37 32 2 13 Dune Messiah (Dune Chronicles #2) 3.86 140 2 106 The Iliad 3.83 1726 2 1371 Harry Potter Collection (Harry Potter, #1-6) 4.73 6 3 10 The Virtue of Selfishness: A New Concept of Egoism 3.47 53 5 665 Sherlock Holmes: The Complete Novels and Stories, Volume I 4.45 13 5 3581 The War of Art: Break Through the Blocks & Win Your Inner Creative Battles 4.07 29 6 1319 Harry Potter Collection (Harry Potter, #1-6) 4.73 6 7 10 The Pilgrimage 3.64 171 8 4004 Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 4.77 11 9 8 Children of Dune (Dune Chronicles #3) 3.9 116 9 112 Icy Sparks 3.68 27 9 3476 Harry Potter and the Goblet of Fire (Harry Potter, #4) 4.53 332 10 6 Atlas Shrugged 3.68 147 10 662 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 12 1 Sherlock Holmes: The Complete Novels and Stories, Volume I 4.45 13 14 3581 The Lover 3.76 148 15 275 The Good Soldier 3.72 362 20 7628 The Thorn Birds 4.22 201 26 3412 The Complete Sherlock Holmes, Vol 2 4.46 117 26 3586 The Dark Tower (The Dark Tower, #7) 4.27 105 26 5091 The Lord of the Rings: Weapons and Warfare 4.53 12 29 36 Slouching Towards Bethlehem 4.24 34 29 424 Atlas Shrugged 3.68 147 29 662 The Elegant Universe: Superstrings, Hidden Dimensions, and the Quest for the Ultimate Theory 4.06 75 29 771 The Known World 3.82 42 30 67 Tropic of Cancer 3.71 205 30 249 The Odyssey 3.73 1703 30 1381

## **QUESTION 4:** Exploring INNER Joins with the **books** and **to\_read** tables

You will need to uncomment the Query to answer the question.

**Part A:**  What do an inner join do?

> <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">Your Answer: Select all rows from both the tables as long as condition is satisfied.</span>

<span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;"><b>Part B:</b>&nbsp;Currently this table has an <b>'IS NOT NULL'</b> in the <b>WHERE</b> clause.&nbsp; Is this necessary for this table?&nbsp; Why or Why not?&nbsp;</span> 

> Your Answer: **'IS NOT NULL'** in the **WHERE** clause is not needed for this query because we are using inner join and there are no null values in book\_titles.

**Part C:**  When using SQL, you are usually returning much larger tables so every line of code will add to the compilation and return time.  Rewrite this inner join query so that it contains no redundant code.   

> Your Answer:  Updated the query in the code box below.

In [9]:

-- INNER JOIN
SELECT TOP 30 b.title, b.average_rating, b.books_count,tr.user_id, tr.book_id
FROM BooksDB.dbo.books AS b
INNER JOIN BooksDB.dbo.to_read AS tr
ON b.book_id = tr.book_id
ORDER BY tr.book_id;

(30 rows affected)

Total execution time: 00:00:00.429

title average_rating books_count user_id book_id Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 12 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 192 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 202 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 244 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 253 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 350 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 478 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 511 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 525 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 527 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 552 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 644 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 754 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 826 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1043 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1205 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1442 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1470 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1712 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1814 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 1919 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2051 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2052 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2158 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2160 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2345 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2476 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2523 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2623 1 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 4.54 275 2673 1

## **QUESTION 5:**  Using joins to find the most tagged book and the most popular tag

We will be writing 3 queries to find both the most tagged book and that book's most popular tag.

**Part A:**  What is the most tagged book?

> <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">Start by joining the </span> **books** <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">and </span> **book\_tags** <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">&nbsp;tables ON the books.best_book_id and book_tags.goodreads_book_id.&nbsp;&nbsp;<br></span> <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">We want the <i>most popular</i> book, so think about ordering the table in a way that will display both the book title and the number of times a book tag has been used.</span>
> 
> _MInimum Desired output:_  The title of the most tagged book, and the number of times the book has been tagged.
> 
> **Title:** **Selected Poems**

**Part B:** How many different tags have been used for the most tagged book?

> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">Start by joining the&nbsp;</span> **books** <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">and&nbsp;</span> **book\_tags** <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">&nbsp;tables ON the books.best_book_id and book_tags.goodreads_book_id.&nbsp;&nbsp;<br></span> We want to find the _most popular_ tag id, so think about ordering in a way that will display the most popular tag id.
> 
> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><i>Minimum Desired Output:</i> The tag id for the most popular tag used for the most tagged book.&nbsp;</span> 
> 
> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><b>Tag Id: 249</b></span>

<span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><b>Part C: </b>What is this particular tag?</span>

> You may return this answer with a simple query.

> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><i>Minimum Desired Output:</i>&nbsp;The tag name for the most popular tag used for the most tagged book.&nbsp;</span> 
> 
> <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);"><b>Tag Name:&nbsp;</b></span> **30574**

In [55]:
-- Code Here:
SELECT TOP (1) b.title, COUNT(tg.tag_id) as "No of time book was tagged"
FROM BooksDB.dbo.books AS b
INNER JOIN BooksDB.dbo.book_tags as tg
ON b.best_book_id = tg.goodreads_book_id
GROUP BY b.title
ORDER BY COUNT(tg.tag_id) DESC;

SELECT COUNT(DISTINCT(tg.tag_id)) as "No of tags tagged for most popular book"
FROM BooksDB.dbo.books AS b
INNER JOIN BooksDB.dbo.book_tags as tg
ON b.best_book_id = tg.goodreads_book_id
WHERE b.title = 'Selected Poems';

SELECT b.title, tg.tag_id, tg.count, COUNT(tg.tag_id) as "No of time book was tagged"
FROM BooksDB.dbo.books AS b
INNER JOIN BooksDB.dbo.book_tags as tg
ON b.best_book_id = tg.goodreads_book_id
WHERE b.title = 'Selected Poems'
GROUP BY title, tg.tag_id, tg.count
ORDER BY tg.count DESC;

(1 row affected)

(1 row affected)

(376 rows affected)

Total execution time: 00:00:00.547

title No of time book was tagged Selected Poems 400

No of tags tagged for most popular book 249

title tag_id count No of time book was tagged Selected Poems 30574 14899 1 Selected Poems 30574 10599 1 Selected Poems 30574 8752 1 Selected Poems 30574 7406 1 Selected Poems 8717 6315 1 Selected Poems 23831 1010 1 Selected Poems 23831 931 1 Selected Poems 23831 929 1 Selected Poems 23831 660 1 Selected Poems 8717 222 1 Selected Poems 7457 201 1 Selected Poems 11557 144 1 Selected Poems 7457 134 1 Selected Poems 8717 129 1 Selected Poems 8717 86 1 Selected Poems 7457 85 1 Selected Poems 11557 77 1 Selected Poems 11557 72 1 Selected Poems 5207 69 1 Selected Poems 20811 67 1 Selected Poems 22743 64 1 Selected Poems 11557 54 1 Selected Poems 22743 41 1 Selected Poems 5207 37 1 Selected Poems 509 32 1 Selected Poems 7457 31 1 Selected Poems 9221 30 1 Selected Poems 23821 28 1 Selected Poems 22743 27 1 Selected Poems 10210 26 1 Selected Poems 23816 25 1 Selected Poems 11590 23 1 Selected Poems 7404 22 1 Selected Poems 18367 22 1 Selected Poems 21689 22 1 Selected Poems 22743 22 1 Selected Poems 22753 22 1 Selected Poems 26211 22 1 Selected Poems 2277 21 1 Selected Poems 509 20 1 Selected Poems 2313 20 1 Selected Poems 7404 20 1 Selected Poems 11743 20 1 Selected Poems 30521 20 1 Selected Poems 5751 19 1 Selected Poems 23816 18 1 Selected Poems 18367 17 1 Selected Poems 20849 17 1 Selected Poems 2277 16 2 Selected Poems 5207 16 1 Selected Poems 23816 16 1 Selected Poems 10644 14 1 Selected Poems 11743 14 1 Selected Poems 14764 14 1 Selected Poems 923 13 1 Selected Poems 20774 13 1 Selected Poems 22753 13 1 Selected Poems 5207 12 1 Selected Poems 10644 12 1 Selected Poems 11590 12 1 Selected Poems 18045 12 1 Selected Poems 18367 12 1 Selected Poems 20318 12 1 Selected Poems 20849 12 1 Selected Poems 923 11 1 Selected Poems 7404 11 1 Selected Poems 11743 11 1 Selected Poems 23860 11 1 Selected Poems 30521 11 1 Selected Poems 15169 10 1 Selected Poems 18045 10 1 Selected Poems 21773 10 1 Selected Poems 2312 9 1 Selected Poems 14764 9 1 Selected Poems 18045 9 1 Selected Poems 21689 9 1 Selected Poems 26735 9 1 Selected Poems 30521 9 1 Selected Poems 30781 9 1 Selected Poems 2313 8 1 Selected Poems 5776 8 1 Selected Poems 9221 8 1 Selected Poems 10641 8 1 Selected Poems 17213 8 1 Selected Poems 20849 8 1 Selected Poems 23229 8 1 Selected Poems 23838 8 1 Selected Poems 30781 8 1 Selected Poems 9221 7 1 Selected Poems 10094 7 1 Selected Poems 10658 7 1 Selected Poems 11590 7 1 Selected Poems 23816 7 1 Selected Poems 25152 7 1 Selected Poems 30521 7 1 Selected Poems 32586 7 1 Selected Poems 32686 7 1 Selected Poems 2312 6 2 Selected Poems 2938 6 1 Selected Poems 7404 6 1 Selected Poems 9221 6 1 Selected Poems 10197 6 1 Selected Poems 11644 6 1 Selected Poems 20318 6 1 Selected Poems 22689 6 2 Selected Poems 23821 6 1 Selected Poems 23824 6 1 Selected Poems 25167 6 1 Selected Poems 30781 6 1 Selected Poems 32724 6 1 Selected Poems 2104 5 1 Selected Poems 5090 5 1 Selected Poems 5775 5 1 Selected Poems 11491 5 1 Selected Poems 11743 5 1 Selected Poems 18367 5 1 Selected Poems 20731 5 1 Selected Poems 20888 5 1 Selected Poems 21689 5 1 Selected Poems 22344 5 1 Selected Poems 22753 5 1 Selected Poems 23040 5 1 Selected Poems 23812 5 1 Selected Poems 23860 5 1 Selected Poems 25562 5 1 Selected Poems 31505 5 1 Selected Poems 31825 5 1 Selected Poems 32315 5 1 Selected Poems 32586 5 1 Selected Poems 134 4 1 Selected Poems 783 4 1 Selected Poems 920 4 1 Selected Poems 1128 4 1 Selected Poems 2271 4 1 Selected Poems 2282 4 1 Selected Poems 2313 4 1 Selected Poems 5393 4 1 Selected Poems 5746 4 1 Selected Poems 6041 4 1 Selected Poems 7423 4 1 Selected Poems 7472 4 1 Selected Poems 7654 4 1 Selected Poems 7658 4 1 Selected Poems 10093 4 1 Selected Poems 10644 4 1 Selected Poems 15169 4 2 Selected Poems 18318 4 2 Selected Poems 20320 4 1 Selected Poems 20731 4 1 Selected Poems 22722 4 1 Selected Poems 23040 4 1 Selected Poems 23349 4 1 Selected Poems 23812 4 1 Selected Poems 23822 4 1 Selected Poems 23853 4 1 Selected Poems 23857 4 1 Selected Poems 23

## **QUESTION 6:** Find a book based on a tag of your choice using joins.

Some suggestions: any hobbies you like? baking, drawing, etc. how about favorite holidays, subjects, animals, etc?

**Part A:** Explore the tags table for a tag of your choice.  Suggested ideas, try searching for one of your hobbies, a favorite time of year, favorite animal, etc.  Start by querying the tag table for your chosen tag.  (_Hint:_ in order to make sure you find all combinations of your tags, use the wildcard).  If your query returns more than one result, pick your favorite.

> > _Here is an example.  Alyce wanted to search for halloween books.  She used the wildcard, '%halloween%' and 8 different combinations were found.  Alyce selected '**halloween-storytime**', with a tag id of 13877._

> _Minimum Desired Output:_ <span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; background-color: rgba(127, 127, 127, 0.1);">tag name and tag id</span>

> **Your Tag and Tag Id: Tag id is 26735 and Tag name is school for book A christmas carol.**

**Part B:** Now that you have the **tag id**, find the title that was tagged the most with this tag by joining the books and books\_tags tables.

> > _Alyce joined the book table with the tags table and discovered that <u>The Little Old Lady Who Was Not Afraid of Anything</u> was tagged as '**halloween-storytime**' the most._

> _Minimum Desired Output:_ <span style="background-color: rgba(127, 127, 127, 0.1); font-family: -apple-system, BlinkMacSystemFont, sans-serif;">title and tag count</span>

> **Your Title and Number of times the tag was used:  Title: The stone soup and 3 : Number of times tagged,**

In [64]:
-- Code Here:
SELECT book_id, title
FROM BooksDB.dbo.books
WHERE title like '%christmas%';

SELECT t.tag_id, b.tag_name
FROM BooksDB.dbo.book_tags as t 
INNER JOIN BooksDB.dbo.tags as b
ON t.tag_id = b.tag_id
WHERE t.goodreads_book_id = '5326';

SELECT  b.title, COUNT(tg.tag_id)
FROM BooksDB.dbo.books AS b
INNER JOIN BooksDB.dbo.book_tags as tg
ON b.best_book_id = tg.goodreads_book_id
WHERE tg.tag_id = '26735'
GROUP BY b.title
ORDER BY COUNT(tg.tag_id) DESC;

(19 rows affected)

(100 rows affected)

(1913 rows affected)

Total execution time: 00:00:01.653

book_id title 5326 A Christmas Carol 113946 How the Grinch Stole Christmas! 196970 The Night Before Christmas 169848 A Christmas Carol, The Chimes and The Cricket on the Hearth 5357 Skipping Christmas 42172 The Christmas Shoes (Christmas Hope #1) 74731 The Stupidest Angel: A Heartwarming Tale of Christmas Terror (Pine Cove, #3) 5338 A Christmas Carol and Other Christmas Writings 239810 The Best Christmas Pageant Ever (The Herdmans #1) 161361 The Christmas Box (The Christmas Box, #1) 16349 Hercule Poirot's Christmas (Hercule Poirot, #20) 13097901 Merry Christmas, Alex Cross (Alex Cross, #19) 3268926 The Christmas Sweater 10844125 The Christmas Wedding 1275105 A Charlie Brown Christmas 6517407 The Christmas List 126185 The Christmas Train 199532 A Redbird Christmas 10518110 The Bite Before Christmas (Argeneau, #15.5; Night Huntress, #6.5)

tag_id tag_name 30574 to-read 7457 classics 8717 currently-reading 11743 fiction 11557 favorites 7404 classic 7224 christmas 11305 fantasy 5207 books-i-own 22743 owned 18367 literature 17213 kindle 14487 historical-fiction 509 19th-century 14694 holiday 3389 audiobook 7423 classic-literature 4949 book-club 26735 school 5751 british 6622 charles-dickens 240 1001-books 22753 owned-books 11590 favourites 3392 audiobooks 22034 novels 26256 rory-gilmore-reading-challenge 9468 dickens 10197 ebook 14467 historical 5776 british-literature 10644 english 31859 victorian 13150 ghosts 10210 ebooks 10641 england 3371 audio 1642 adult 6953 childrens 25114 read-for-school 6857 children 6828 childhood 24960 re-read 3358 audible 27669 short-stories 236 1001 7417 classic-fiction 1659 adult-fiction 18045 library 10658 english-literature 25152 read-in-2016 25151 read-in-2015 7422 classic-lit 9221 default 26251 rory-gilmore-challenge 25150 read-in-2014 21989 novel 12315 for-school 26740 school-books 251 1001-books-to-read-before-you-die 22973 paranormal 33114 young-adult 20849 my-library 30521 to-buy 20774 my-books 10094 e-books 14704 holidays 7478 classics-to-read 3379 audio-books 18326 literary-fiction 12948 general-fiction 7228 christmas-books 26246 rory-gilmore 10093 e-book 25148 read-in-2012 15169 i-own 20811 my-ebooks 17217 kindle-books 25167 read-in-english 5746 brit-lit 5775 british-lit 29011 supernatural 25149 read-in-2013 22025 novella 9886 drama 5090 books 25147 read-in-2011 2104 all-time-favorites 29716 the-classics 25765 reread 3377 audio-book 30464 time-travel 6888 children-s 31430 uk 25234 read-more-than-once 26257 rory-gilmore-reading-list 6895 children-s-books 254 1001-books-you-must-read-before-you 30635 to-read-classics 24999 read-2016

title (No column name) Stone Soup 3 Selected Poems 2 The Collected Poems 2 Leviathan 2 One Flew Over the Cuckoo's Nest 2 The Complete Stories 2 The Complete Stories and Poems 1 The Complete Works 1 The Compound (The Compound, #1) 1 The Constitution of the United States of America 1 The Cost of Discipleship 1 The Count of Monte Cristo 1 The Cricket in Times Square 1 The Cross and the Switchblade 1 The Cross of Lead (Crispin, #1) 1 The Crossover 1 The Crucible 1 The Crying of Lot 49 1 The Curious Case of Benjamin Button 1 The Curious Incident of the Dog in the Night-Time 1 The Dark Is Rising (The Dark is Rising, #2) 1 The Daughter of Time (Inspector Alan Grant, #5) 1 The Day Jimmy's Boa Ate the Wash 1 The Day of the Locust 1 The Day the Crayons Quit 1 The Dead 1 The Dead and the Gone (Last Survivors, #2) 1 The Death of Ivan Ilych 1 The Decameron 1 The Declaration of Independence and The Constitution of the United States 1 The Design of Everyday Things 1 The Devil's Arithmetic 1 The Diary of a Young Girl 1 The Dinner 1 The Disappearing Spoon: And Other True Tales of Madness, Love, and the History of the World from the Periodic Table of the Elements 1 The Disreputable History of Frankie Landau-Banks 1 The Divine Comedy 1 The Diving Bell and the Butterfly 1 The Dot 1 The Double Helix 1 The Ear, the Eye, and the Arm 1 The Earth, My Butt, and Other Big Round Things 1 The Edible Woman 1 The Education of Little Tree 1 The Egypt Game (Game, #1) 1 The Element: How Finding Your Passion Changes Everything 1 The Elements of Style 1 The Emperor's Code (The 39 Clues, #8) 1 The End of Poverty 1 The End of the Affair 1 The English Patient 1 The Epic of Gilgamesh 1 The Evolution of Calpurnia Tate (Calpurnia Tate, #1) 1 The Face on the Milk Carton (Janie Johnson, #1) 1 The Faerie Queene 1 The Fairy-Tale Detectives (The Sisters Grimm, #1) 1 The Fall 1 The Fall of the House of Usher 1 The Fall of the House of Usher and Other Tales 1 The Famished Road 1 The Federalist Papers 1 The Fifth Discipline: The Art & Practice of The Learning Organization 1 The Fire Next Time 1 The Fire Within (The Last Dragon Chronicles, #1) 1 The First Four Years (Little House, #9) 1 The First Part Last (Heaven, #2) 1 The Five Chinese Brothers 1 The Five Dysfunctions of a Team: A Leadership Fable 1 The Five People You Meet in Heaven 1 The Folk of the Faraway Tree (The Faraway Tree, #3) 1 The Foot Book: Dr. Seuss's Wacky Book of Opposites (Bright and Early Board Books) 1 The Fountainhead 1 The Fourteenth Goldfish 1 The Freedom Writers Diary 1 The French Lieutenant's Woman 1 The Frog and Toad Treasury: Frog and Toad are Friends/Frog and Toad Together/Frog and Toad All Year (Frog and Toad #1-3) 1 The Future of Us 1 The Gay Science 1 The Gentlemen's Alliance †, Vol. 1 1 The Ghost Map: The Story of London's Most Terrifying Epidemic - and How It Changed Science, Cities, and the Modern World 1 The Gift of the Magi 1 The Gift of Therapy: An Open Letter to a New Generation of Therapists and Their Patients 1 The Giver (The Giver, #1) 1 The Giving Tree 1 The Glass Castle 1 The Glass Menagerie 1 The Goal: A Process of Ongoing Improvement 1 The God of Small Things 1 The Good Earth (House of Earth, #1) 1 The Good Soldier 1 The Great Brain (Great Brain #1) 1 The Great Divorce 1 The Great Gatsby 1 The Great Gilly Hopkins 1 The Greatest Generation 1 The Grouchy Ladybug 1 The Gruffalo 1 The Guns of Navarone 1 The Handmaid's Tale 1 The Happy Prince 1 The Hat 1 The Haunted Mask (Goosebumps, #11) 1 The Haunting of Hill House 1 The Heart is a Lonely Hunter 1 The Heart of the Matter 1 The Hero With a Thousand Faces 1 The Hiding Place: The Triumphant True Story of Corrie Ten Boom 1 The High Window (Philip Marlowe, #3) 1 The Histories 1 The History of Sexuality, Volume 1: An Introduction 1 The History of Tom Jones, a Foundling 1 The Hobbit 1 The Hobbit: Graphic Novel 1 The Hot Zone: The Terrifying True Story of the Origins of the Ebola Virus 1 The Hound of the Baskervilles 1 The Hours 1 The House a

## **QUESTION 7:** What are the top 10 most popular book titles on the 'to read' list?

**Part A:**  Explore how the **to\_read** and **books** tables are organized, and then join them.  What will you use as a key?

> Your Answer: book\_id

**Part B:** Organize your newly joined table in a way that groups popularity based on the title based on users.

> _Hint:_ Suggest using an alias for the aggregation you will need to run on the **user\_id**

> _Minimum Desired Output:_ title and aliased column

In [82]:
-- Code Here:

SELECT b.title, b.book_id, tr.user_id 
FROM BooksDB.dbo.to_read as tr
INNER JOIN BooksDB.dbo.books AS b
ON tr.book_id = b.book_id ;


SELECT b.title, COUNT(tr.user_id) AS "No of users"
FROM BooksDB.dbo.to_read as tr
INNER JOIN BooksDB.dbo.books AS b
ON tr.book_id = b.book_id 
GROUP BY b.title
ORDER BY "No of users" DESC;





(87493 rows affected)

(811 rows affected)

Displaying Top 5000 rows.

Total execution time: 00:00:03.014

title book_id user_id Children of Dune (Dune Chronicles #3) 112 1 The Hitchhiker's Guide to the Galaxy (Hitchhiker's Guide to the Galaxy, #1) 11 2 The Ultimate Hitchhiker's Guide to the Galaxy 13 2 Dune Messiah (Dune Chronicles #2) 106 2 The Iliad 1371 2 Harry Potter Collection (Harry Potter, #1-6) 10 3 The Virtue of Selfishness: A New Concept of Egoism 665 5 Sherlock Holmes: The Complete Novels and Stories, Volume I 3581 5 The War of Art: Break Through the Blocks & Win Your Inner Creative Battles 1319 6 Harry Potter Collection (Harry Potter, #1-6) 10 7 The Pilgrimage 4004 8 Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 8 9 Children of Dune (Dune Chronicles #3) 112 9 Icy Sparks 3476 9 Harry Potter and the Goblet of Fire (Harry Potter, #4) 6 10 Atlas Shrugged 662 10 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 1 12 Sherlock Holmes: The Complete Novels and Stories, Volume I 3581 14 The Lover 275 15 The Good Soldier 7628 20 The Thorn Birds 3412 26 The Complete Sherlock Holmes, Vol 2 3586 26 The Dark Tower (The Dark Tower, #7) 5091 26 The Lord of the Rings: Weapons and Warfare 36 29 Slouching Towards Bethlehem 424 29 Atlas Shrugged 662 29 The Elegant Universe: Superstrings, Hidden Dimensions, and the Quest for the Ultimate Theory 771 29 The Known World 67 30 Tropic of Cancer 249 30 The Odyssey 1381 30 The Pilgrimage 4004 30 The Hobbit 5907 30 The Hitchhiker's Guide to the Galaxy (Hitchhiker's Guide to the Galaxy, #1) 11 32 The Richest Man in Babylon 1052 33 Inés of My Soul 3300 33 Lysistrata 1591 34 The Omnivore's Dilemma: A Natural History of Four Meals 3109 34 Sherlock Holmes: The Complete Novels and Stories, Volume I 3581 34 Heidi 93 39 Heretics of Dune (Dune Chronicles #5) 117 39 Time Enough for Love 353 39 Another Bullshit Night in Suck City 386 39 Libra 400 39 The Virtue of Selfishness: A New Concept of Egoism 665 39 Hamlet 1420 39 Chapterhouse: Dune (Dune Chronicles #6) 105 46 The Lord of the Rings: The Art of The Fellowship of the Ring 119 46 The Elegant Universe: Superstrings, Hidden Dimensions, and the Quest for the Ultimate Theory 771 46 The Confusion (The Baroque Cycle, #2) 822 46 Tunnels of Blood (Cirque Du Freak, #3) 8957 47 Cryptonomicon 816 51 Mornings on Horseback 2368 54 One Fish, Two Fish, Red Fish, Blue Fish 7770 54 The Broken Wings 291 55 Memoirs of a Geisha 930 55 The Taste of Home Cookbook 3885 58 The Thorn Birds 3412 60 Sherlock Holmes: The Complete Novels and Stories, Volume I 3581 60 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 3 61 The Ultimate Hitchhiker's Guide to the Galaxy 13 61 The Lord of the Rings: Weapons and Warfare 36 61 The Lord of the Rings: Weapons and Warfare 36 64 The Power of One (The Power of One, #1) 122 65 The Door Into Summer 348 67 Job: A Comedy of Justice 355 67 The Salmon of Doubt (Dirk Gently, #3) 359 67 Cryptonomicon 816 67 The Odyssey 1381 67 The Men Who Stare at Goats 1824 67 The World Is Flat: A Brief History of the Twenty-first Century 1911 67 The Jane Austen Book Club 2152 67 The Tipping Point: How Little Things Can Make a Big Difference 2612 67 A Little Princess 3008 67 The Corrections 3805 67 Regeneration (Regeneration, #1) 5872 67 The Virtue of Selfishness: A New Concept of Egoism 665 69 Persuasion 2156 69 A Prayer for Owen Meany 4473 69 The Adventures of Sherlock Holmes 3590 71 Children of Dune (Dune Chronicles #3) 112 72 Memoirs of a Geisha 930 72 Cryptonomicon 816 75 Think and Grow Rich: The Landmark Bestseller - Now Revised and Updated for the 21st Century 1005 75 Sherlock Holmes: The Complete Novels and Stories, Volume I 3581 78 What to Expect the First Year (What to Expect) 98 82 Tropic of Cancer 249 82 Trace (Kay Scarpetta, #13) 6530 82 Cryptonomicon 816 84 Harry Potter Collection (Harry Potter, #1-6) 10 94 The Lord of the Rings: Weapons and Warfare 36 98 Charlie and the Chocolate Factory (Charlie Bucket, #1) 6310 99 Children of Dune (Dune Chronicles #3) 112 101 Digging to America 698 104 Persuasion 2156 105 A Home at the End of th

title No of users The Ultimate Hitchhiker's Guide to the Galaxy 1812 The Hitchhiker's Guide to the Galaxy (Hitchhiker's Guide to the Galaxy, #1) 1767 The Lord of the Rings: The Art of The Fellowship of the Ring 1499 Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5) 1498 Harry Potter and the Goblet of Fire (Harry Potter, #4) 1484 The Known World 1352 Harry Potter and the Prisoner of Azkaban (Harry Potter, #3) 1293 The Lord of the Rings: Weapons and Warfare 1211 Notes from a Small Island 1148 Harry Potter Collection (Harry Potter, #1-6) 1110 The Lord of the Rings (The Lord of the Rings, #1-3) 1094 The Salmon of Doubt (Dirk Gently, #3) 1069 Dune Messiah (Dune Chronicles #2) 1051 Treasure Island 1040 Tropic of Cancer 1007 Harry Potter and the Half-Blood Prince (Harry Potter, #6) 973 Children of Dune (Dune Chronicles #3) 913 The Power of One (The Power of One, #1) 866 Heretics of Dune (Dune Chronicles #5) 815 J.R.R. Tolkien 4-Book Boxed Set: The Hobbit and The Lord of the Rings 801 Job: A Comedy of Justice 737 The Door Into Summer 724 Mostly Harmless (Hitchhiker's Guide to the Galaxy, #5) 696 I am Charlotte Simmons 661 The Broken Wings 659 Killing Yourself to Live: 85% of a True Story 608 The Portrait of a Lady 580 The Lover 567 Slouching Towards Bethlehem 564 We the Living 563 Cryptonomicon 551 Fullmetal Alchemist, Vol. 1 (Fullmetal Alchemist, #1) 533 Libra 525 Maurice 525 The Elegant Universe: Superstrings, Hidden Dimensions, and the Quest for the Ultimate Theory 524 Chapterhouse: Dune (Dune Chronicles #6) 513 One Hundred Years of Solitude 504 What to Expect the First Year (What to Expect) 495 I'm a Stranger Here Myself: Notes on Returning to America after Twenty Years Away 484 The Fellowship of the Ring (The Lord of the Rings, #1) 450 The Lost Continent: Travels in Small Town America 447 Everyday Italian: 125 Simple and Delicious Recipes 445 Heidi 439 Dirk Gently's Holistic Detective Agency (Dirk Gently #1) 438 Tropic of Capricorn 438 Neither Here nor There: Travels in Europe 438 City of Glass (The New York Trilogy, #1) 435 In a Sunburned Country 431 The Virtue of Selfishness: A New Concept of Egoism 430 A Short History of Nearly Everything 427 Harry Potter and the Order of the Phoenix (Harry Potter, #5) 400 Sherlock Holmes: The Complete Novels and Stories, Volume I 387 Daniel Deronda 385 Madame Bovary 384 Stranger in a Strange Land 374 Digging to America 373 The Da Vinci Code (Robert Langdon, #2) 372 The Mother Tongue: English and How It Got That Way 348 The Phantom Tollbooth 346 The Long Dark Tea-Time of the Soul (Dirk Gently, #2) 343 The Game: Penetrating the Secret Society of Pickup Artists 319 Breaking the Spell: Religion as a Natural Phenomenon 316 The Tipping Point: How Little Things Can Make a Big Difference 315 Robinson Crusoe 309 Agamemnon (Oresteia, #1) 306 Collapse: How Societies Choose to Fail or Succeed 306 Trump: The Art of the Deal 305 Harry Potter and the Sorcerer's Stone (Harry Potter, #1) 287 Fire Sea (The Death Gate Cycle, #3) 286 Wild Swans: Three Daughters of China 279 Metamorphoses 277 Think and Grow Rich: The Landmark Bestseller - Now Revised and Updated for the 21st Century 273 Hatchet (Brian's Saga, #1) 269 Middlesex 268 The Confusion (The Baroque Cycle, #2) 267 Hamlet 265 Atlas Shrugged 262 Snow Crash 261 Pompeii 260 The Odyssey 250 Leviathan 249 The Iliad/The Odyssey 240 Zodiac 240 The Diamond Age: or, A Young Lady's Illustrated Primer 239 Fahrenheit 451 236 Play It as It Lays 232 The Iliad 220 Corelli's Mandolin 219 The Brooklyn Follies 217 Gravity's Rainbow 217 Memoirs of a Geisha 213 Time Enough for Love 208 Moon Palace 207 The Design of Everyday Things 206 The Restaurant at the End of the Universe (Hitchhiker's Guide, #2) 202 Wild Fire (John Corey, #4) 198 Nickel and Dimed: On (Not) Getting By in America 195 Falling Angels 194 Martin Chuzzlewit 189 Giada's Family Dinners 189 Wild at Heart: Discovering the Secret of a Man's Soul 184 The New York Trilogy 182 Surely You're Joking, Mr. Feynman!: 

## **QUESTION 8:** Largest 'To Read' Lists

**Part A:**  Create a query that returns a table that contains the top 10 users with the most number of titles on their 'to read' list.    Group all the entries by user\_id.  

> _Hint:_ You might want an alias for the titles.
> 
> _Minimum Desired Output:_  user\_id and your aliased column of titles.

**Part B:**  The longest list length is 15 titles total, and is shared by 4 different users.  Select one of the uses and print out their entire '**to read**' list.

In [88]:
-- Code Here:
SELECT TOP (10) tr.user_id, count(b.title) AS "No of books"
FROM BooksDB.dbo.to_read as tr
INNER JOIN BooksDB.dbo.books AS b
ON tr.book_id = b.book_id
GROUP BY tr.user_id
ORDER BY COUNT(b.title) DESC;

SELECT tr.user_id, b.title 
FROM BooksDB.dbo.to_read as tr
INNER JOIN BooksDB.dbo.books AS b
ON tr.book_id = b.book_id
WHERE tr.user_id = 12483;


(10 rows affected)

(15 rows affected)

Total execution time: 00:00:01.196

user_id No of books 14771 15 12483 15 11932 15 38201 15 36923 14 41523 14 27933 14 37359 14 25734 13 51137 13

user_id title 12483 Fahrenheit 451 12483 Hamlet 12483 Great Expectations 12483 Madame Bovary 12483 A Room with a View 12483 Cryptonomicon 12483 The Complete Sherlock Holmes, Vol 2 12483 Tropic of Cancer 12483 The New York Trilogy 12483 Metamorphoses 12483 Founding Brothers: The Revolutionary Generation 12483 Peter and the Shadow Thieves (Peter and the Starcatchers, #2) 12483 The Salmon of Doubt (Dirk Gently, #3) 12483 Jailbird 12483 The Broken Wings